src/libstd/sys/unix/fs.rs

```diff
@@ -251,21 +251,35 @@ impl Iterator for ReadDir {
 
     #[cfg(not(any(target_os = "solaris", target_os = "fuchsia")))]
     fn next(&mut self) -> Option<io::Result<DirEntry>> {
-        unsafe {
-            let mut ret = DirEntry {
-                entry: mem::zeroed(),
-                root: self.root.clone()
-            };
-            let mut entry_ptr = ptr::null_mut();
-            loop {
-                if readdir64_r(self.dirp.0, &mut ret.entry, &mut entry_ptr) != 0 {
-                    return Some(Err(Error::last_os_error()))
-                }
-                if entry_ptr.is_null() {
-                    return None
-                }
-                if ret.name_bytes() != b"." && ret.name_bytes() != b".." {
-                    return Some(Ok(ret))
+        if self.dirp.is_none() {
+            return None;
+        }
+
+        match self.dirp {
+            None => return None,
+            ref mut dirp @ Some(_) => {
+                unsafe {
+                    let mut ret = DirEntry {
+                        entry: mem::zeroed(),
+                        root: self.root.clone()
+                    };
+                    let mut entry_ptr = ptr::null_mut();
+                    loop {
+                        if readdir64_r(dirp.as_mut().unwrap().0.as_mut(),
+                                       &mut ret.entry,
+                                       &mut entry_ptr) != 0 {
+                            if entry_ptr.is_null() {
```

Comment:
This is the actual change being made, right?

Reply:
Yes. If `readdir64_r` returns a non-empty exit code *and* signals the end of the directory stream by writing `nullptr` to `entry_ptr`, we still want to return the OS error in this iteration but `None` on the next iteration.

Reply:
I think this change can be made a bit simpler by just using a null pointer instead of the whole option business. Then the diff just becomes a `if self.dirp.0.is_null() { return None; }` at the top of Next, and an `if entry_ptr.is_null() { self.dirp.0 = ptr::null_mut(); }` in the error case here.